runner/docker: integrate ComfyUI LivePortrait

emranemran opened the pull request.

leszko commented:
@emranemran I've built the ComfyUI image, but I'm not sure how to use it 🙃 

I see some samples in https://github.com/kijai/ComfyUI-LivePortraitKJ/tree/main/examples, I think the most closest to what we do is this one: https://github.com/kijai/ComfyUI-LivePortraitKJ/blob/main/examples/liveportrait_realtime_example_01.json

But still this workflow is quite big, could you try to covert it into something that we can use in our runner, like one of these ones? https://www.notion.so/livepeer/Video-Processing-ComfyUI-Nodes-Workflows-1340a34856878005b0fcf9eb02d1819a?pvs=4#1580a348568780f1a718fbf49bf98d4c

emranemran commented:
@leszko some notes for you: 
I used the [realtime one](https://github.com/kijai/ComfyUI-LivePortraitKJ/blob/main/examples/liveportrait_realtime_example_01.json) as a reference and made the following changes: 
 - the source images will be generated via the `LoadImageFromUrlOrPath` node which I added to the comfyui docker file. This allows us to pass an url in the json provided by the user instead of having to load from disk. 
 - replace the webcam capture pieces with just `LoadImage` node which will be driven from our pipeline (comfystream call). 
 - TODO: the example uses the `ImageResizeKJ` and `GetImageSizeandCount` nodes. I'm not 100% sure if these are needed but I tried adding them via the repo you see in the comfyui docker file - howevere they don't seem to be detected when I run comfyui locally: 
![Screen Shot 2024-12-12 at 1 01 10 AM](https://github.com/user-attachments/assets/b57a4d39-28fb-47d4-82a6-0b1fd84c46da)

So the previous note I shared re: comfystream supporting only a single LoadImage node can (should?) be handled with the above modified workflow if we can figure out the missing dependency. In other words, the condition [here](https://github.com/yondonfu/comfystream/blob/main/src/comfystream/utils.py#L16) and [here](https://github.com/yondonfu/comfystream/blob/main/src/comfystream/utils.py#L38) should still be valid and we won't need to modify comfystream for the time being (discussed with Yondon as well). 

**Next Steps**
* address the TODO above -- i've attached the json files generated by the UI here
[liveportrait_live_example.json](https://github.com/user-attachments/files/18108590/liveportrait_live_example.json)
* then integrate the json into the pipeline

emranemran commented:
One more note @leszko: i think the PreviewImage node should be replaced with the SaveImage node so that the img is returned to the pipeline. I forgot to update this in the example above.

emranemran commented:
One more update @leszko: I was staring at the comfyui logs and noticed this: 
```
ModuleNotFoundError: No module named 'model_management'

Cannot import /Users/emran/custom_nodes/ComfyUI-KJNodes module for custom nodes: No module named 'model_management'
```
so hopefully installing that module will get us past that missing node error.

leszko commented on runner/docker/Dockerfile.live-base-comfyui:
```diff
@@ -34,6 +34,23 @@ RUN cd /comfyui/custom_nodes && \
+    cd ComfyUI-KJNodes && \
+    pip install -r requirements.txt
+# Install ComfyUI-load-image-from-url Node (https://github.com/tsogzark/ComfyUI-load-image-from-url.git)
+RUN cd /comfyui/custom_nodes && \
```
This does not build for me for some reason. But I haven't investigated yet why.

leszko commented on runner/docker/Dockerfile.live-base-comfyui:
```diff
@@ -34,6 +34,23 @@ RUN cd /comfyui/custom_nodes && \
+# Install ComfyUI-load-image-from-url Node (https://github.com/tsogzark/ComfyUI-load-image-from-url.git)
+RUN cd /comfyui/custom_nodes && \
+    git clone https://github.com/tsogzark/ComfyUI-load-image-from-url.git && \
+    cd ComfyUI-load-image-from-url && \
+    pip install -r requirements.txt
```
```suggestion
    cd ComfyUI-load-image-from-url
```

leszko commented:
@emranemran I spent some time today on your PR. Added some inline comments. Apart from that:

**1. Error "No module named 'model_management'"**
I was not able to solve it. Seems like a missing dependency, but I was not able to install it. Instead I removed the custom node `ComfyUI-KJNodes`. IMO we may actually not need it. 

**2. Running E2E**
I've created the JSON Workflow which I believe should work, but I get some weird error which I don't understand. Maybe it's just some type or maybe I'm doing something completely wrong. Maybe you'll find it helpful (attaching as JSON to this comment + included it in the draft PR 👇 

Error that I get:
```
2024-12-12 14:55:35 ERROR    Error in process run method: {"type": "invalid_prompt", "message": "Cannot execute because node DownloadAndLoadLivePortraitModels does not exist.", "details": "Node ID '#3'", "extra_info": {}}
```

I've created a draft PR to you branch (the intention is not merge it but rather to present what I'm talking about in points 1 and 2.

[liveportrait.json](https://github.com/user-attachments/files/18113699/liveportrait.json)

leszko commented:
> @emranemran I spent some time today on your PR. Added some inline comments. Apart from that:
> 
> **1. Error "No module named 'model_management'"** I was not able to solve it. Seems like a missing dependency, but I was not able to install it. Instead I removed the custom node `ComfyUI-KJNodes`. IMO we may actually not need it.
> 
> **2. Running E2E** I've created the JSON Workflow which I believe should work, but I get some weird error which I don't understand. Maybe it's just some type or maybe I'm doing something completely wrong. Maybe you'll find it helpful (attaching as JSON to this comment + included it in the draft PR 👇
> 
> Error that I get:
> 
> ```
> 2024-12-12 14:55:35 ERROR    Error in process run method: {"type": "invalid_prompt", "message": "Cannot execute because node DownloadAndLoadLivePortraitModels does not exist.", "details": "Node ID '#3'", "extra_info": {}}
> ```
> 
> I've created a draft PR to you branch (the intention is not merge it but rather to present what I'm talking about in points 1 and 2. #359
> 
> [liveportrait.json](https://github.com/user-attachments/files/18113699/liveportrait.json)

emranemran commented on runner/dl_checkpoints.sh:
```diff
@@ -92,6 +92,7 @@ function download_live_models() {
     huggingface-cli download yuvraj108c/Depth-Anything-Onnx --include depth_anything_vitl14.onnx --local-dir models/ComfyUI--models/Depth-Anything-Onnx
     download_sam2_checkpoints
     download_stable_diffusion_checkpoints
+    huggingface-cli download Kijai/LivePortrait_safetensors --local-dir models/ComfyUI--models/livePortrait
```
hmm is this suggestion correct? i see we do a `RUN ln -s /models/ComfyUI--models /comfyui/models` in the dockerfile and LivePortrait expects the files to be in `/comfyui/models/liveportrait`. It seems to be able to load the models from here without this suggestion.

leszko replied:
Yeah, I think your're right

emranemran commented:
@leszko i'm seeing the same `No face detected in FIRST source image` error when testing with the local trickle method and using an input driving video with human faces on it. Here are the tests I ran today: 

* My first step was to try and get a local workflow running on my mac. See attached workflow: (
[liveportrait-works-local-mac-export-for-docker.json](https://github.com/user-attachments/files/18122931/liveportrait-works-local-mac-export-for-docker.json) ) which *did* work and running Queue generates frames within ~12s on my mac m1. Feel free to try it on your end. 
* Using this as a reference, I modified it to use the Load/SaveTensor nodes and replaced the webcam node with LoadImage -- see my `DEBUG` commit on this PR. Using this workflow, I still get the same `No face detected in FIRST source image` error. 
* So I'm thinking there's two options here: 
  - the Issue list on the LivePortrait github repo has similar complaints from some users. One of the suggestion was to have a very neutral face when starting the pipeline otherwise it's finicky and won't work correctly. 
  - the frame being passed to the workflow is corrupted. This isn't true because I saved the frames to disk to verify and I see the frames as expected with the faces (see DEBUG commit as well). 
  
So next steps: 
  * The server trickle testing method doesn't work on my end - my router/firewal drops packets to mediamtx ingress so rest of pipeline is not reliable. Could you try it out with the changes in this PR and see if it just happens to work?
  * This is not a great idea because it might introduce unintended effects but maybe we modify this logic to try multiple frames instead of just 1? https://github.com/kijai/ComfyUI-LivePortraitKJ/blob/main/liveportrait/live_portrait_pipeline.py#L77

leszko reviewed (approved):
I got it working with the following JSON Workflow:
```json
{
        "1": {
            "inputs": {
                "precision": "auto",
                "mode": "human"
            },
            "class_type": "DownloadAndLoadLivePortraitModels",
            "_meta": {
                "title": "(Down)Load LivePortraitModels"
            }
        },
        "189": {
            "inputs": {
                "dsize": 512,
                "scale": 2.34,
                "vx_ratio": 0.099,
                "vy_ratio": 0.148,
                "face_index": 0,
                "face_index_order": "large-small",
                "rotate": false,
                "pipeline": [
                    "1",
                    0
                ],
                "cropper": [
                    "204",
                    0
                ],
                "source_image": [
                    "210",
                    0
                ]
            },
            "class_type": "LivePortraitCropper",
            "_meta": {
                "title": "LivePortrait Cropper"
            }
        },
        "190": {
            "inputs": {
                "lip_zero": false,
                "lip_zero_threshold": 0.03,
                "stitching": true,
                "delta_multiplier": 1,
                "mismatch_method": "constant",
                "relative_motion_mode": "single_frame",
                "driving_smooth_observation_variance": 0.000003,
                "expression_friendly": false,
                "expression_friendly_multiplier": 1,
                "pipeline": [
                    "1",
                    0
                ],
                "crop_info": [
                    "189",
                    1
                ],
                "source_image": [
                    "210",
                    0
                ],
                "driving_images": [
                    "196",
                    0
                ]
            },
            "class_type": "LivePortraitProcess",
            "_meta": {
                "title": "LivePortrait Process"
            }
        },
        "196": {
            "inputs": {},
            "class_type": "LoadTensor",
            "_meta": {
                "title": "LoadTensor"
            }
        },
        "204": {
            "inputs": {
                "landmarkrunner_onnx_device": "torch_gpu",
                "keep_model_loaded": true
            },
            "class_type": "LivePortraitLoadMediaPipeCropper",
            "_meta": {
                "title": "LivePortrait Load MediaPipeCropper"
            }
        },
        "210": {
            "inputs": {
                "url_or_path": "https://raw.githubusercontent.com/kijai/ComfyUI-LivePortraitKJ/refs/heads/main/assets/examples/source/s7.jpg"
            },
            "class_type": "LoadImageFromUrlOrPath",
            "_meta": {
                "title": "LoadImageFromUrlOrPath"
            }
        },
        "211": {
            "inputs": {
                "images": [
                    "190",
                    0
                ]
            },
            "class_type": "SaveTensor",
            "_meta": {
                "title": "SaveTensor"
            }
        }
    }
```
